adapter/src/test/java/com/twilio/kudu/sql/JDBCQueryIT.java

```diff
@@ -278,6 +278,25 @@ public void testTopNQuery() throws Exception {
     }
   }
 
+  @Test
+  public void testLimit() throws Exception {
+    try (Connection conn = DriverManager.getConnection(JDBC_URL)) {
+      String sqlFormat = "SELECT * FROM \"ReportCenter.DeliveredMessages\" WHERE account_sid = " + "'%s' LIMIT 2";
+      String sql = String.format(sqlFormat, JDBCQueryIT.ACCOUNT_SID);
+      String expectedPlan = "KuduToEnumerableRel\n" + "  KuduLimitRel(limit=[2])\n"
+          + "    KuduFilterRel(ScanToken 1=[account_sid EQUAL AC1234567])\n"
+          + "      KuduQuery(table=[[kudu, ReportCenter.DeliveredMessages]])\n";
```

Comment:
Yes, instead of reading all the rows that pass the filter we check each time when the sortedKeyPrefix changes to see if we have read enough grouped rows and stop scanning if we have. The returned rows and then sorted on the client. 
For a query that orders by time descending and groups by timestamp truncated to the hour we will read all rows from the latest hour , check if we have read enough grouped rows if yes we stop scanning, if no we read all rows from the previous hour and so on.